tests/http/server/_base.py

```diff
@@ -98,3 +111,413 @@ def _test_disconnect(
+            0, f"Running _test_cancellation_at_every_await for {test_name}..."
+        )
+
+        for request_number in itertools.count(1):
+            (
+                Deferred___next__,
+                unblock_awaits,
+                get_awaits_seen,
+                has_seen_new_await,
+            ) = self.create_deferred___next___patch(seen_awaits, request_number)
+
+            try:
+                with mock.patch(
+                    "synapse.http.server.respond_with_json", wraps=respond_with_json
+                ) as respond_mock:
+                    with mock.patch.object(
+                        Deferred,
+                        "__next__",
+                        new=Deferred___next__,
+                    ):
+                        # Start the request.
+                        channel = make_request()
+                        request = channel.request
+
+                        if request_number == 0:
+                            self.assertFalse(
+                                respond_mock.called,
+                                "Request finished before we could disconnect - "
+                                "was `await_result=False` passed to `make_request`?",
+                            )
+                        else:
+                            # Requests after the first may be lucky enough to hit caches
+                            # all the way through and never have to block.
+                            pass
+
+                        # Run the request until we see a new `await` which we have not
+                        # yet cancelled at, or it completes.
+                        while not respond_mock.called and not has_seen_new_await():
+                            previous_awaits_seen = get_awaits_seen()
+
+                            reactor.pump([0.0])
+
+                            if get_awaits_seen() == previous_awaits_seen:
+                                # We didn't see any progress. Try advancing the clock.
+                                reactor.pump([1.0])
+
+                            if get_awaits_seen() == previous_awaits_seen:
+                                # We still didn't see any progress. The request might be stuck.
+                                self.fail(
+                                    "Request appears to be stuck, possibly due to "
+                                    "a previous cancelled request"
+                                )
+
+                    if respond_mock.called:
+                        # The request ran to completion and we are done with testing it.
+
+                        # `respond_with_json` writes the response asynchronously, so we
+                        # might have to give the reactor a kick before the channel gets
+                        # the response.
+                        reactor.pump([1.0])
+
+                        self.assertEqual(channel.code, expected_code)
+                        return channel.json_body
```

Comment:
can we just `return channel` and leave these up to the caller?

Reply:
Possibly? There may be an unfortunate interaction where we block the `await`s in `_async_write_json_to_request_in_thread`.
I'll have to test it.